tslearn/nonmyopic.py

```diff
@@ -0,0 +1,509 @@
+    --------------------
+    >>>
+
+    References
+    --------------------
+    [1] A. Dachraoui, A. Bondu & A. Cornuéjols. Early classification of time series as a non myopic sequential decision
+    making problem. 2015 Conference paper
+
+
+    """
+
+    def __init__(
+        self,
+        number_cluster,
+        solver,
+        hidden_layer_sizes,
+        random_state,
+        maximum_iteration,
+        minimum_time_stamp,
+        lamb,
+        cost_time_parameter,
+    ):
+        super(NonMyopicEarlyClassification, self).__init__()
+        self.solver = solver
+        self.hidden_layer_sizes = hidden_layer_sizes
+        self.random_state = random_state
+        self.maximum_iteration = maximum_iteration
+        self.number_cluster = number_cluster
+        self.minimum_time_stamp = minimum_time_stamp
+        self.lamb = lamb
+        self.cost_time_parameter = cost_time_parameter
+
+    def clustering(self, X):
+        """
+        This function compute and fit clusters for a training dataset of time series using K-means and euclidean
+        distances
+        Parameters
+        ----------
+        X: array-like
+            The training dataset of time series
+
+        Returns
+        -------
+        Vector : a vector as long as the number of series in the training dataset that indicates in which cluster the
+        series at the same index belongs
+
+         Float : the silhouette score of the fitting of the clustering
+        """
+        self.cluster_ = TimeSeriesKMeans(n_clusters=self.number_cluster)
+        c_k = self.cluster_.fit_predict((to_time_series_dataset(X)))
+        silhouette = silhouette_score(
+            to_time_series_dataset(X), c_k, metric="euclidean"
+        )
+
+        return c_k, silhouette
+
+    def fit(self, X, Y):
+        """
+        This function fits classifiers that are currently multilayer perceptrons to a training set of time series and
+        associated classes. A classifier is fit for each time stamp above a minimum time stamp that is an attribute of
+        the class of the model. Then some probabilities are computed using clusters already computed.
+        This function should be divided among 2 or 3 functions.
+
+        Parameters
+        ----------
+        X: Array-like
+            a dataset of time series
+        Y: vector
+            the associated classes of the series from X_train
+        """
+        classes_y = np.unique(Y)
+        c_k, silhouette = self.clustering(X)
+        self.silhouette_ = silhouette
+        self.classifier_ = []
+        self.clusters_ = {}
+        self.pyhatyck_ = {}
+        self.indice_ck_ = []
+        self.number_classes_ = len(classes_y)
+        self.len_X_ = X.shape[1]
+        mid_X = int(X.shape[0] / 2)
+        X1 = X[:mid_X]
+        Y1 = Y[:mid_X]
+        X2 = X[mid_X:]
+        Y2 = Y[mid_X:]
+        c_k2 = c_k[mid_X:]
+        self.pyck_ = np.zeros(shape=(self.number_classes_, self.number_cluster))
```

Comment:
pyck_ can be constructed using `scipy.sparse.coo_matrix` as done for `confusion_matrix` in scikit-learn.

https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_matrix.html

One important thing here is the note: 
> By default when converting to CSR or CSC format, duplicate (i,j) entries will be summed together. This facilitates efficient construction of finite element matrices and the like. (see example)

so you could have something like:

```python
weights = np.ones((self.n_classes_, self.n_clusters)) / (self.n_classes_ * self.n_clusters)
self.pyck_ = coo_matrix((weights, (y_classes, y_clusters)),
                    shape=(self.n_classes_, self.n_clusters), dtype=np.float32,
                    ).toarray()
```

Reply:
self.pyck_ is now computed using coo_matrix function but it did not remove the for loop just made it lighter.

Reply:
Also this means that the classes must be expressed with the labels {0, 1, 2,...} and not for instance {-1, 1} or otherwise.

Reply:
Well, that means you probably need a dictionary somewhere that maps actual class labels (that could even be strings, I think, in the sklearn context) to indices 0, 1, 2.

Something like:

```python 
self.labels_ = sorted(set(y_classes))
y_classes_indices = [self.labels_.index(yi) for yi in y_classes]
```